### PR TITLE
table.vstack shouldn't modify inputs

### LINKED ISSUE #16119: table.vstack alters inputs when Rows
### Description

When `astropy.table.vstack` is given a list of `Row` objects, it modifies the input list to become a list of `Table` objects instead.

### Expected behavior

`vstack` should not alter its inputs at all.  The input list of Rows should remain a list of Rows and not become a list of Table objects instead.

### How to Reproduce

Environment created with:
```
conda create -n test python=3.10
source activate test
pip install astropy==6.0.0
```

Example failure
```python
from astropy.table import Table, vstack
t1 = Table(data=dict(x=[1,2,3], y=['a', 'b', 'c']))
rows = [r for r in t1]
print(f"{type(rows[0])=}")   # ... <class 'astropy.table.row.Row'>
print(f"First row x is {rows[0]['x']:.1f}")   # 1.0
t2 = vstack(rows)  # unexpectedly alters input list of Rows to become list of Tables
print(f"{type(rows[0])=}")   # ... <class 'astropy.table.table.Table'>
print(f"First row x is {rows[0]['x']:.1f}")  # crashes with TypeError: unsupported format string passed to Column.__format__
```

### Versions

```python
import platform; print(platform.platform())
import sys; print("Python", sys.version)
import astropy; print("astropy", astropy.__version__)
import numpy; print("Numpy", numpy.__version__)
import erfa; print("pyerfa", erfa.__version__)
try:
    import scipy
    print("Scipy", scipy.__version__)
except ImportError:
    print("Scipy not installed")
try:
    import matplotlib
    print("Matplotlib", matplotlib.__version__)
except ImportError:
    print("Matplotlib not installed")
```
```
macOS-10.16-x86_64-i386-64bit
Python 3.10.13 (main, Sep 11 2023, 08:39:02) [Clang 14.0.6 ]
astropy 6.0.0
Numpy 1.26.4
pyerfa 2.0.1.1
Scipy not installed
Matplotlib not installed
```

### PR DESCRIPTION
**Pull request by sbailey**
### Description

This PR updates `table.vstack` so that it won't modify the input list,
including a test demonstrating the original problem (it now passes).
`table.vstack` still supports inputs being lists of Tables or Rows, but it no longer modifies the original user-provided list in-place, thus avoiding an unexpected side-effect of `vstack`.  See #16119 for context.

Fixes #16119

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Comment by pllim** [comment]
pre-commit.ci autofix

**Comment by pllim** [comment]
Hmm the autofix didn't fix these:

```
astropy/table/operations.py:65:14: C416 Unnecessary `list` comprehension (rewrite using `list()`)
astropy/table/tests/test_operations.py:1547:16: C416 Unnecessary `list` comprehension (rewrite using `list()`)
```

File: astropy/table/operations.py
Comment (mhvk) [inline astropy/table/operations.py:65]: This is a nice catch! I'm also fine with it, but I think the code would be clearer if we just constructed a new list as we went, like
```
result = []
for val in tables:
    if isinstance(val, Table):
        pass  # short-circuit common case
    elif isinstance(val, Row):
        val = Table(val)
    elif ...
    ...
    result.append(val)

return result
```
What do you think?

File: astropy/table/tests/test_operations.py
Comment (mhvk) [inline astropy/table/tests/test_operations.py:1548]: I'd make this a bit stronger:
```
rows0 = rows[0]
table.vstack(rows)
assert rows[0] is rows0
```
I don't think you need the rest of the test (or write it for a `Quantity`, but really fine without too).

File: docs/changes/table/16130.bugfix.rst
Comment (mhvk) [inline docs/changes/table/16130.bugfix.rst:2]: strictly: "contains non-Table objects like ``astropy.table.Row``"

**Review by mhvk — COMMENTED** [review]
Thanks for not just reporting the bug but also finding the cause and fixing it!

Some somewhat nitpicky comments - with the main one being to clean up the code a bit while you are at it.

File: astropy/table/operations.py
Comment (sbailey) [reply astropy/table/operations.py:65]: It's admittedly a matter of taste.  While I agree that your suggestion is correct and guarantees that the inputs aren't modified, I also tend to be suspicious of code that assembles lists with `list.append` in `if/elif/then` chains because it risks future bugs with a mis-placed `continue` statement or some logic that otherwise accidentally misses the `append`.

But I will defer to you as an astropy maintainer to make the style call, and update the code if you feel strongly enough about it (I don't feel strongly about it).  You should have write access to the source branch.

File: astropy/table/tests/test_operations.py
Comment (sbailey) [reply astropy/table/tests/test_operations.py]: Good idea; I strengthened the test with the `assert rows[0] is rows0` check.  I think it is still useful to do the null test that Tables remain Tables too, so I left that (while also following the `tables[0] is tables0` pattern).  I didn't add `Quantity` testing, since I'm not familiar with stacking those into a Table.

File: docs/changes/table/16130.bugfix.rst
Comment (sbailey) [reply docs/changes/table/16130.bugfix.rst]: Good clarification.  Done.

**Review by mhvk — APPROVED** [review]
Thanks for making the change to the test. I don't feel strongly enough about my other suggestion to change it, so let's just get this in!

**Comment by pllim** [comment]
For future reference, please use "Squash and Merge" when the PR commit history is messy. Now the merge commit in this PR history is in `main`. Not the end of the world but it is something we avoid. FYI.

**Comment by mhvk** [comment]
Darn, I'm normally more careful! I fear my poor eyes may have seen the 7 as a 1...

**Pull request merged**
